packages/sdk/src/protocols/store.ts

```diff
@@ -0,0 +1,317 @@
+    for await (const page of this.queryGenerator(decoders, options)) {
+      const _promises = page.map(async (msgPromise) => {
+        if (abort) return;
+        abort = Boolean(await callback(msgPromise));
+      });
+
+      await Promise.all(_promises);
+      if (abort) break;
+    }
+  }
+
+  createCursor(message: IDecodedMessage): Cursor {
+    if (
+      !message ||
+      !message.timestamp ||
+      !message.payload ||
+      !message.contentTopic
+    ) {
+      throw new Error("Message is missing required fields");
+    }
+
+    const contentTopicBytes = utf8ToBytes(message.contentTopic);
+
+    const digest = sha256(concat([contentTopicBytes, message.payload]));
+
+    const messageTime = BigInt(message.timestamp.getTime()) * BigInt(1000000);
+
+    return {
+      digest,
+      pubsubTopic: message.pubsubTopic,
+      senderTime: messageTime,
+      receiverTime: messageTime
+    };
+  }
+
+  private validateDecodersAndPubsubTopic<T extends IDecodedMessage>(
+    decoders: IDecoder<T>[],
+    options?: waku_store.QueryOptions
+  ): {
+    pubsubTopic: string;
+    contentTopics: string[];
+    decodersAsMap: Map<string, IDecoder<T>>;
+  } {
+    if (decoders.length === 0) {
+      throw new Error("No decoders provided");
+    }
+
+    // convert array to set to remove duplicates
+    const uniquePubsubTopicsInQuery = Array.from(
+      new Set(decoders.map((decoder) => decoder.pubsubTopic))
+    );
+    // If multiple pubsub topics are provided, throw an error
+    if (uniquePubsubTopicsInQuery.length > 1) {
+      throw new Error(
+        "API does not support querying multiple pubsub topics at once"
+      );
+    }
+
+    // we can be certain that there is only one pubsub topic in the query
+    const pubsubTopicForQuery = uniquePubsubTopicsInQuery[0];
+
+    ensurePubsubTopicIsConfigured(
+      pubsubTopicForQuery,
+      this.protocol.pubsubTopics
+    );
+
+    // check that the pubsubTopic from the Cursor and Decoder match
+    if (
+      options?.cursor?.pubsubTopic &&
+      options.cursor.pubsubTopic !== pubsubTopicForQuery
+    ) {
+      throw new Error(
+        `Cursor pubsub topic (${options?.cursor?.pubsubTopic}) does not match decoder pubsub topic (${pubsubTopicForQuery})`
+      );
+    }
+
+    const decodersAsMap = new Map();
+    decoders.forEach((dec) => {
+      if (decodersAsMap.has(dec.contentTopic)) {
+        throw new Error(
+          "API does not support different decoder per content topic"
+        );
+      }
+      decodersAsMap.set(dec.contentTopic, dec);
+    });
+
+    const contentTopics = decoders
+      .filter((decoder) => decoder.pubsubTopic === pubsubTopicForQuery)
+      .map((dec) => dec.contentTopic);
+
+    if (contentTopics.length === 0) {
+      throw new Error("No decoders found for topic " + pubsubTopicForQuery);
+    }
+
+    return {
+      pubsubTopic: pubsubTopicForQuery,
+      contentTopics,
+      decodersAsMap
+    };
+  }
+
+  private constructOptions(
```

Comment:
it seems this function is added where PR's name implies only movement, I would prefer for this kind of changes one PR for just moving code and another with improvements

Reply:
just to highlight: no `additional` logic was added part of this PR:
`constructOptions` was created as an abstraction to move pieces of code into a function to clean the original function.

Reply:
I understand and not against it - proposing in future having one PR with moving and consecutive PR with changes noticed

this way if we have to do cherry pick or revert of something granular in master - it will be easier

Reply:
@danisharora099 , @adklempner do you have any thoughts or preferences on this?

Reply:
i agree we should aim at keeping the diff minimal, and have the diff correspond to the PR's purpose as closely as possible